File: src/treemapper/diffctx/edges/config/kubernetes.py
from __future__ import annotations

import re
from collections import defaultdict
from pathlib import Path

from ...types import Fragment, FragmentId
from ..base import EdgeBuilder, EdgeDict

_K8S_API_VERSION_RE = re.compile(r"^apiVersion:\s?([^\s#]{1,100})", re.MULTILINE)
_K8S_KIND_RE = re.compile(r"^kind:\s?(\w{1,100})", re.MULTILINE)
_K8S_NAME_RE = re.compile(r"^\s{1,20}name:\s?['\"]?([^'\"#\n]{1,200})", re.MULTILINE)
_K8S_NAMESPACE_RE = re.compile(r"^\s{1,20}namespace:\s?['\"]?([^'\"#\n]{1,200})", re.MULTILINE)

_CONFIGMAP_REF_RE = re.compile(r"configMapKeyRef:\s?\n\s{1,20}name:\s?['\"]?([^'\"#\n]{1,200})", re.MULTILINE)
_CONFIGMAP_NAME_RE = re.compile(r"configMapName:\s?['\"]?([^'\"#\n]{1,200})", re.MULTILINE)
_SECRET_REF_RE = re.compile(r"secretKeyRef:\s?\n\s{1,20}name:\s?['\"]?([^'\"#\n]{1,200})", re.MULTILINE)
_SECRET_NAME_RE = re.compile(r"secretName:\s?['\"]?([^'\"#\n]{1,200})", re.MULTILINE)

_SERVICE_NAME_RE = re.compile(r"serviceName:\s?['\"]?([^'\"#\n]{1,200})", re.MULTILINE)
_BACKEND_SERVICE_RE = re.compile(r"service:\s?\n\s{1,20}name:\s?['\"]?([^'\"#\n]{1,200})", re.MULTILINE)

_IMAGE_RE = re.compile(r"^\s{1,20}image:\s?['\"]?([^'\"#\n]{1,300})", re.MULTILINE)

_SELECTOR_MATCH_LABELS_RE = re.compile(
    r"selector:\s?\n\s{1,20}matchLabels:\s?\n((?:\s{1,20}[a-zA-Z0-9_./-]{1,100}:\s?[^\n:]{1,200}\n){1,50})", re.MULTILINE
)
_LABELS_RE = re.compile(r"labels:\s?\n((?:\s{1,20}[a-zA-Z0-9_./-]{1,100}:\s?[^\n:]{1,200}\n){1,50})", re.MULTILINE)
_LABEL_PAIR_RE = re.compile(r"^\s{0,20}([a-zA-Z0-9_./-]{1,100}):\s?['\"]?([a-zA-Z0-9_./-]{1,100})['\"]?\s{0,10}$", re.MULTILINE)
_SIMPLE_SELECTOR_RE = re.compile(r"selector:\s?\n((?:\s{1,20}[a-zA-Z0-9_./-]{1,100}:\s?[^\n:]{1,200}\n){1,50})", re.MULTILINE)

_VOLUME_CONFIGMAP_RE = re.compile(r"configMap:\s?\n\s{1,20}name:\s?['\"]?([^'\"#\n]{1,200})", re.MULTILINE)
_VOLUME_SECRET_RE = re.compile(r"secret:\s?\n\s{1,20}secretName:\s?['\"]?([^'\"#\n]{1,200})", re.MULTILINE)
_VOLUME_PVC_RE = re.compile(r"persistentVolumeClaim:\s?\n\s{1,20}claimName:\s?['\"]?([^'\"#\n]{1,200})", re.MULTILINE)

_YAML_EXTS = {".yaml", ".yml"}

_K8S_KINDS = {
    "Deployment",
    "Service",
    "ConfigMap",
    "Secret",
    "Ingress",
    "Pod",
    "ReplicaSet",
    "StatefulSet",
    "DaemonSet",
    "Job",
    "CronJob",
    "PersistentVolume",
    "PersistentVolumeClaim",
    "ServiceAccount",
    "Role",
    "RoleBinding",
    "ClusterRole",
    "ClusterRoleBinding",
    "NetworkPolicy",
    "HorizontalPodAutoscaler",
    "Namespace",
}


def _is_kubernetes_manifest(path: Path, content: str | None = None) -> bool:
    if path.suffix.lower() not in _YAML_EXTS:
        return False

    if content is None:
        try:
            content = path.read_text(encoding="utf-8")
        except (OSError, UnicodeDecodeError):
            return False

    api_match = _K8S_API_VERSION_RE.search(content)
    kind_match = _K8S_KIND_RE.search(content)

    if api_match and kind_match:
        kind = kind_match.group(1).strip()
        return kind in _K8S_KINDS

    return False


def _extract_resource_info(content: str) -> tuple[str | None, str | None, str | None]:
    kind_match = _K8S_KIND_RE.search(content)
    name_match = _K8S_NAME_RE.search(content)
    namespace_match = _K8S_NAMESPACE_RE.search(content)

    kind = kind_match.group(1).strip() if kind_match else None
    name = name_match.group(1).strip() if name_match else None
    namespace = namespace_match.group(1).strip() if namespace_match else None

    return kind, name, namespace


def _extract_label_pairs(label_block: str) -> dict[str, str]:
    return {m.group(1).strip(): m.group(2).strip() for m in _LABEL_PAIR_RE.finditer(label_block)}


def _extract_labels_by_pattern(content: str, pattern: re.Pattern[str]) -> dict[str, str]:
    labels: dict[str, str] = {}
    for match in pattern.finditer(content):
        labels.update(_extract_label_pairs(match.group(1)))
    return labels


def _extract_labels(content: str) -> dict[str, str]:
    return _extract_labels_by_pattern(content, _LABELS_RE)


def _extract_selector_labels(content: str) -> dict[str, str]:
    return _extract_labels_by_pattern(content, _SELECTOR_MATCH_LABELS_RE)


def _labels_match(selector: dict[str, str], labels: dict[str, str]) -> bool:
    if not selector:
        return False
    return all(labels.get(k) == v for k, v in selector.items())


def _find_k8s_files(changed_files: list[Path]) -> list[Path]:
    k8s_files: list[Path] = []
    for f in changed_files:
        if f.suffix.lower() not in _YAML_EXTS:
            continue
        try:
            content = f.read_text(encoding="utf-8")
            if _is_kubernetes_manifest(f, content):
                k8s_files.append(f)
        except (OSError, UnicodeDecodeError):
            continue
    return k8s_files


def _collect_k8s_dirs(k8s_files: list[Path]) -> set[Path]:
    k8s_dirs: set[Path] = set()
    for f in k8s_files:
        k8s_dirs.add(f.parent)
        if f.parent.name in {"base", "overlays", "templates", "manifests"}:
            k8s_dirs.add(f.parent.parent)
    return k8s_dirs


def _is_in_k8s_dir(candidate: Path, k8s_dirs: set[Path]) -> bool:
    for k8s_dir in k8s_dirs:
        try:
            if candidate.is_relative_to(k8s_dir):
                return True
        except (ValueError, TypeError):
            continue
    return False


class _K8sIndex:
    configmaps: dict[str, list[FragmentId]]
    secrets: dict[str, list[FragmentId]]
    services: dict[str, list[FragmentId]]
    pvcs: dict[str, list[FragmentId]]
    pods_with_labels: list[tuple[FragmentId, dict[str, str]]]
    images: dict[str, list[FragmentId]]

    def __init__(self) -> None:
        self.configmaps = defaultdict(list)
        self.secrets = defaultdict(list)
        self.services = defaultdict(list)
        self.pvcs = defaultdict(list)
        self.pods_with_labels = []
        self.images = defaultdict(list)


class KubernetesEdgeBuilder(EdgeBuilder):
    weight = 0.65
    configmap_secret_weight = 0.70
    service_weight = 0.60
    selector_weight = 0.55
    image_weight = 0.40
    reverse_weight_factor = 0.45

    def discover_related_files(
        self,
        changed_files: list[Path],
        all_candidate_files: list[Path],
        repo_root: Path | None = None,
    ) -> list[Path]:
        k8s_files = _find_k8s_files(changed_files)
        if not k8s_files:
            return []

        k8s_dirs = _collect_k8s_dirs(k8s_files)
        changed_set = set(changed_files)
        discovered: list[Path] = []

        for candidate in all_candidate_files:
            if candidate in changed_set or candidate.suffix.lower() not in _YAML_EXTS:
                continue
            if not _is_in_k8s_dir(candidate, k8s_dirs):
                continue
            try:
                content = candidate.read_text(encoding="utf-8")
                if _is_kubernetes_manifest(candidate, content):
                    discovered.append(candidate)
            except (OSError, UnicodeDecodeError):
                continue

        return discovered

    def build(self, fragments: list[Fragment], repo_root: Path | None = None) -> EdgeDict:
        k8s_fragments = [f for f in fragments if _is_kubernetes_manifest(f.path, f.content)]

        if not k8s_fragments:
            return {}

        edges: EdgeDict = {}
        idx = self._build_resource_index(k8s_fragments)

        for frag in k8s_fragments:
            self._build_configmap_edges(frag, idx.configmaps, edges)
            self._build_secret_edges(frag, idx.secrets, edges)
            self._build_service_edges(frag, idx.services, edges)
            self._build_volume_edges(frag, idx.pvcs, edges)
            self._build_selector_edges(frag, idx.pods_with_labels, edges)
            self._build_image_edges(frag, idx.images, edges)

        return edges

    def _build_resource_index(self, k8s_fragments: list[Fragment]) -> _K8sIndex:
        idx = _K8sIndex()
        for frag in k8s_fragments:
            self._index_fragment(frag, idx)
        return idx

    def _index_fragment(self, frag: Fragment, idx: _K8sIndex) -> None:
        kind, name, _ = _extract_resource_info(frag.content)

        if name:
            self._index_by_kind(kind, name, frag.id, idx)

        if kind in {"Pod", "Deployment", "StatefulSet", "DaemonSet", "ReplicaSet", "Job", "CronJob"}:
            labels = _extract_labels(frag.content)
            if labels:
                idx.pods_with_labels.append((frag.id, labels))

        self._index_images(frag, idx)

    def _index_by_kind(self, kind: str | None, name: str, frag_id: FragmentId, idx: _K8sIndex) -> None:
        if kind == "ConfigMap":
            idx.configmaps[name].append(frag_id)
        elif kind == "Secret":
            idx.secrets[name].append(frag_id)
        elif kind == "Service":
            idx.services[name].append(frag_id)
        elif kind == "PersistentVolumeClaim":
            idx.pvcs[name].append(frag_id)

    def _index_images(self, frag: Fragment, idx: _K8sIndex) -> None:
        for match in _IMAGE_RE.finditer(frag.content):
            image = match.group(1).strip()
            if image and not image.startswith("$"):
                idx.images[image].append(frag.id)

    def _link_by_patterns(
        self,
        frag: Fragment,
        patterns: list[re.Pattern[str]],
        index: dict[str, list[FragmentId]],
        edges: EdgeDict,
        weight: float,
    ) -> None:
        for pattern in patterns:
            for match in pattern.finditer(frag.content):
                name = match.group(1).strip()
                for target_id in index.get(name, []):
                    if target_id != frag.id:
                        self.add_edge(edges, frag.id, target_id, weight)

    def _build_configmap_edges(
        self,
        frag: Fragment,
        configmaps: dict[str, list[FragmentId]],
        edges: EdgeDict,
    ) -> None:
        patterns = [_CONFIGMAP_REF_RE, _CONFIGMAP_NAME_RE, _VOLUME_CONFIGMAP_RE]
        self._link_by_patterns(frag, patterns, configmaps, edges, self.configmap_secret_weight)

    def _build_secret_edges(
        self,
        frag: Fragment,
        secrets: dict[str, list[FragmentId]],
        edges: EdgeDict,
    ) -> None:
        patterns = [_SECRET_REF_RE, _SECRET_NAME_RE, _VOLUME_SECRET_RE]
        self._link_by_patterns(frag, patterns, secrets, edges, self.configmap_secret_weight)

    def _build_service_edges(
        self,
        frag: Fragment,
        services: dict[str, list[FragmentId]],
        edges: EdgeDict,
    ) -> None:
        patterns = [_SERVICE_NAME_RE, _BACKEND_SERVICE_RE]
        self._link_by_patterns(frag, patterns, services, edges, self.service_weight)

    def _build_volume_edges(
        self,
        frag: Fragment,
        pvcs: dict[str, list[FragmentId]],
        edges: EdgeDict,
    ) -> None:
        self._link_by_patterns(frag, [_VOLUME_PVC_RE], pvcs, edges, self.weight)

    def _build_selector_edges(
        self,
        frag: Fragment,
        pods_with_labels: list[tuple[FragmentId, dict[str, str]]],
        edges: EdgeDict,
    ) -> None:
        kind, _, _ = _extract_resource_info(frag.content)
        if kind != "Service":
            return

        selector = self._get_service_selector(frag.content)
        if not selector:
            return

        for pod_id, labels in pods_with_labels:
            if pod_id != frag.id and _labels_match(selector, labels):
                self.add_edge(edges, frag.id, pod_id, self.selector_weight)

    def _get_service_selector(self, content: str) -> dict[str, str]:
        selector = _extract_selector_labels(content)
        if selector:
            return selector

        selector_match = _SIMPLE_SELECTOR_RE.search(content)
        if not selector_match:
            return {}

        return _extract_label_pairs(selector_match.group(1))

    def _build_image_edges(
        self,
        frag: Fragment,
        images: dict[str, list[FragmentId]],
        edges: EdgeDict,
    ) -> None:
        for match in _IMAGE_RE.finditer(frag.content):
            image = match.group(1).strip()
            if not image or image.startswith("$"):
                continue
            for other_id in images.get(image, []):
                if other_id != frag.id:
                    self.add_edge(edges, frag.id, other_id, self.image_weight)

File: diffctx/src/edges/config_edges/kubernetes.rs
use std::path::{Path, PathBuf};

use once_cell::sync::Lazy;
use regex::Regex;
use rustc_hash::{FxHashMap, FxHashSet};

use crate::types::{Fragment, FragmentId};

use super::super::EdgeDict;
use super::super::base::{self, EdgeBuilder, add_edge};

const WEIGHT: f64 = 0.65;
const CONFIGMAP_SECRET_WEIGHT: f64 = 0.70;
const SERVICE_WEIGHT: f64 = 0.60;
const SELECTOR_WEIGHT: f64 = 0.55;
const IMAGE_WEIGHT: f64 = 0.40;
const REVERSE_FACTOR: f64 = 0.45;

static YAML_EXTS: Lazy<FxHashSet<&str>> = Lazy::new(|| [".yaml", ".yml"].iter().copied().collect());

static K8S_API_VERSION_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"(?m)^apiVersion:\s?([^\s#]{1,100})").unwrap());
static K8S_KIND_RE: Lazy<Regex> = Lazy::new(|| Regex::new(r"(?m)^kind:\s?(\w{1,100})").unwrap());
static K8S_METADATA_NAME_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r##"(?m)^metadata:\s*\n\s{2,4}name:\s?['"]?([^'"#\n]{1,200})"##).unwrap()
});
static K8S_NAME_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r##"(?m)^\s{1,20}name:\s?['"]?([^'"#\n]{1,200})"##).unwrap());
static CONFIGMAP_REF_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r##"(?m)configMapKeyRef:\s?\n\s{1,20}name:\s?['"]?([^'"#\n]{1,200})"##).unwrap()
});
static CONFIGMAP_NAME_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r##"(?m)configMapName:\s?['"]?([^'"#\n]{1,200})"##).unwrap());
static SECRET_REF_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r##"(?m)secretKeyRef:\s?\n\s{1,20}name:\s?['"]?([^'"#\n]{1,200})"##).unwrap()
});
static SECRET_NAME_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r##"(?m)secretName:\s?['"]?([^'"#\n]{1,200})"##).unwrap());

static SERVICE_NAME_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r##"(?m)serviceName:\s?['"]?([^'"#\n]{1,200})"##).unwrap());
static BACKEND_SERVICE_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r##"(?m)service:\s?\n\s{1,20}name:\s?['"]?([^'"#\n]{1,200})"##).unwrap()
});

static IMAGE_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r##"(?m)^\s{1,20}image:\s?['"]?([^'"#\n]{1,300})"##).unwrap());

static SELECTOR_MATCH_LABELS_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"(?m)selector:\s?\n\s{1,20}matchLabels:\s?\n((?:\s{1,20}[a-zA-Z0-9_./-]{1,100}:\s?[^\n:]{1,200}\n){1,50})")
        .unwrap()
});
static LABELS_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"(?m)labels:\s?\n((?:\s{1,20}[a-zA-Z0-9_./-]{1,100}:\s?[^\n:]{1,200}\n){1,50})")
        .unwrap()
});
static LABEL_PAIR_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(
        r##"(?m)^\s{0,20}([a-zA-Z0-9_./-]{1,100}):\s?['"]?([a-zA-Z0-9_./-]{1,100})['"]?\s{0,10}$"##,
    )
    .unwrap()
});
static SIMPLE_SELECTOR_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"(?m)selector:\s?\n((?:\s{1,20}[a-zA-Z0-9_./-]{1,100}:\s?[^\n:]{1,200}\n){1,50})")
        .unwrap()
});

static VOLUME_CONFIGMAP_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r##"(?m)configMap:\s?\n\s{1,20}name:\s?['"]?([^'"#\n]{1,200})"##).unwrap()
});
static VOLUME_SECRET_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r##"(?m)secret:\s?\n\s{1,20}secretName:\s?['"]?([^'"#\n]{1,200})"##).unwrap()
});
static VOLUME_PVC_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r##"(?m)persistentVolumeClaim:\s?\n\s{1,20}claimName:\s?['"]?([^'"#\n]{1,200})"##)
        .unwrap()
});

static K8S_KINDS: Lazy<FxHashSet<&str>> = Lazy::new(|| {
    [
        "Deployment",
        "Service",
        "ConfigMap",
        "Secret",
        "Ingress",
        "Pod",
        "ReplicaSet",
        "StatefulSet",
        "DaemonSet",
        "Job",
        "CronJob",
        "PersistentVolume",
        "PersistentVolumeClaim",
        "ServiceAccount",
        "Role",
        "RoleBinding",
        "ClusterRole",
        "ClusterRoleBinding",
        "NetworkPolicy",
        "HorizontalPodAutoscaler",
        "Namespace",
    ]
    .iter()
    .copied()
    .collect()
});

static WORKLOAD_KINDS: Lazy<FxHashSet<&str>> = Lazy::new(|| {
    [
        "Pod",
        "Deployment",
        "StatefulSet",
        "DaemonSet",
        "ReplicaSet",
        "Job",
        "CronJob",
    ]
    .iter()
    .copied()
    .collect()
});

fn is_yaml_file(path: &Path) -> bool {
    let ext = base::file_ext(path);
    YAML_EXTS.contains(ext.as_str())
}

fn is_kubernetes_manifest(path: &Path, content: &str) -> bool {
    if !is_yaml_file(path) {
        return false;
    }

    let api_match = K8S_API_VERSION_RE.is_match(content);
    let kind_match = K8S_KIND_RE.captures(content);

    if let (true, Some(cap)) = (api_match, kind_match) {
        let kind = cap[1].trim();
        K8S_KINDS.contains(kind)
    } else {
        false
    }
}

fn extract_resource_info(content: &str) -> (Option<String>, Option<String>) {
    let kind = K8S_KIND_RE
        .captures(content)
        .map(|c| c[1].trim().to_string());

    let name = K8S_METADATA_NAME_RE
        .captures(content)
        .or_else(|| K8S_NAME_RE.captures(content))
        .map(|c| c[1].trim().to_string());

    (kind, name)
}

fn extract_label_pairs(label_block: &str) -> FxHashMap<String, String> {
    let mut pairs = FxHashMap::default();
    for cap in LABEL_PAIR_RE.captures_iter(label_block) {
        pairs.insert(cap[1].trim().to_string(), cap[2].trim().to_string());
    }
    pairs
}

fn extract_labels_by_pattern(content: &str, pattern: &Regex) -> FxHashMap<String, String> {
    let mut labels = FxHashMap::default();
    for m in pattern.captures_iter(content) {
        labels.extend(extract_label_pairs(&m[1]));
    }
    labels
}

fn extract_labels(content: &str) -> FxHashMap<String, String> {
    extract_labels_by_pattern(content, &LABELS_RE)
}

fn extract_selector_labels(content: &str) -> FxHashMap<String, String> {
    extract_labels_by_pattern(content, &SELECTOR_MATCH_LABELS_RE)
}

fn labels_match(selector: &FxHashMap<String, String>, labels: &FxHashMap<String, String>) -> bool {
    if selector.is_empty() {
        return false;
    }
    selector
        .iter()
        .all(|(k, v)| labels.get(k).map_or(false, |lv| lv == v))
}

fn collect_k8s_dirs(k8s_files: &[&PathBuf]) -> FxHashSet<PathBuf> {
    let mut dirs = FxHashSet::default();
    let special_dirs: FxHashSet<&str> = ["base", "overlays", "templates", "manifests"]
        .iter()
        .copied()
        .collect();

    for f in k8s_files {
        if let Some(parent) = f.parent() {
            dirs.insert(parent.to_path_buf());
            if let Some(dir_name) = parent.file_name().and_then(|n| n.to_str()) {
                if special_dirs.contains(dir_name) {
                    if let Some(grandparent) = parent.parent() {
                        dirs.insert(grandparent.to_path_buf());
                    }
                }
            }
        }
    }
    dirs
}

fn is_in_k8s_dir(candidate: &Path, k8s_dirs: &FxHashSet<PathBuf>) -> bool {
    for dir in k8s_dirs {
        if candidate.starts_with(dir) {
            return true;
        }
    }
    false
}

struct K8sIndex {
    configmaps: FxHashMap<String, Vec<FragmentId>>,
    secrets: FxHashMap<String, Vec<FragmentId>>,
    services: FxHashMap<String, Vec<FragmentId>>,
    pvcs: FxHashMap<String, Vec<FragmentId>>,
    pods_with_labels: Vec<(FragmentId, FxHashMap<String, String>)>,
    images: FxHashMap<String, Vec<FragmentId>>,
}

impl K8sIndex {
    fn new() -> Self {
        Self {
            configmaps: FxHashMap::default(),
            secrets: FxHashMap::default(),
            services: FxHashMap::default(),
            pvcs: FxHashMap::default(),
            pods_with_labels: Vec::new(),
            images: FxHashMap::default(),
        }
    }
}

fn index_by_kind(kind: Option<&str>, name: &str, frag_id: &FragmentId, idx: &mut K8sIndex) {
    match kind {
        Some("ConfigMap") => idx
            .configmaps
            .entry(name.to_string())
            .or_default()
            .push(frag_id.clone()),
        Some("Secret") => idx
            .secrets
            .entry(name.to_string())
            .or_default()
            .push(frag_id.clone()),
        Some("Service") => idx
            .services
            .entry(name.to_string())
            .or_default()
            .push(frag_id.clone()),
        Some("PersistentVolumeClaim") => idx
            .pvcs
            .entry(name.to_string())
            .or_default()
            .push(frag_id.clone()),
        _ => {}
    }
}

fn index_images(frag: &Fragment, idx: &mut K8sIndex) {
    for cap in IMAGE_RE.captures_iter(&frag.content) {
        let image = cap[1].trim();
        if !image.is_empty() && !image.starts_with('$') {
            idx.images
                .entry(image.to_string())
                .or_default()
                .push(frag.id.clone());
        }
    }
}

fn index_fragment(frag: &Fragment, idx: &mut K8sIndex) {
    let (kind, name) = extract_resource_info(&frag.content);

    if let Some(ref n) = name {
        index_by_kind(kind.as_deref(), n, &frag.id, idx);
    }

    if let Some(ref k) = kind {
        if WORKLOAD_KINDS.contains(k.as_str()) {
            let labels = extract_labels(&frag.content);
            if !labels.is_empty() {
                idx.pods_with_labels.push((frag.id.clone(), labels));
            }
        }
    }

    index_images(frag, idx);
}

fn build_resource_index(k8s_fragments: &[&Fragment]) -> K8sIndex {
    let mut idx = K8sIndex::new();
    for frag in k8s_fragments {
        index_fragment(frag, &mut idx);
    }
    idx
}

fn link_by_patterns(
    frag: &Fragment,
    patterns: &[&Regex],
    index: &FxHashMap<String, Vec<FragmentId>>,
    edges: &mut EdgeDict,
    weight: f64,
) {
    for pattern in patterns {
        for cap in pattern.captures_iter(&frag.content) {
            let name = cap[1].trim();
            if let Some(target_ids) = index.get(name) {
                for target_id in target_ids {
                    if *target_id != frag.id {
                        add_edge(edges, &frag.id, target_id, weight, REVERSE_FACTOR);
                    }
                }
            }
        }
    }
}

fn build_configmap_edges(
    frag: &Fragment,
    configmaps: &FxHashMap<String, Vec<FragmentId>>,
    edges: &mut EdgeDict,
) {
    link_by_patterns(
        frag,
        &[&CONFIGMAP_REF_RE, &CONFIGMAP_NAME_RE, &VOLUME_CONFIGMAP_RE],
        configmaps,
        edges,
        CONFIGMAP_SECRET_WEIGHT,
    );
}

fn build_secret_edges(
    frag: &Fragment,
    secrets: &FxHashMap<String, Vec<FragmentId>>,
    edges: &mut EdgeDict,
) {
    link_by_patterns(
        frag,
        &[&SECRET_REF_RE, &SECRET_NAME_RE, &VOLUME_SECRET_RE],
        secrets,
        edges,
        CONFIGMAP_SECRET_WEIGHT,
    );
}

fn build_service_edges(
    frag: &Fragment,
    services: &FxHashMap<String, Vec<FragmentId>>,
    edges: &mut EdgeDict,
) {
    link_by_patterns(
        frag,
        &[&SERVICE_NAME_RE, &BACKEND_SERVICE_RE],
        services,
        edges,
        SERVICE_WEIGHT,
    );
}

fn build_volume_edges(
    frag: &Fragment,
    pvcs: &FxHashMap<String, Vec<FragmentId>>,
    edges: &mut EdgeDict,
) {
    link_by_patterns(frag, &[&VOLUME_PVC_RE], pvcs, edges, WEIGHT);
}

fn get_service_selector(content: &str) -> FxHashMap<String, String> {
    let selector = extract_selector_labels(content);
    if !selector.is_empty() {
        return selector;
    }

    match SIMPLE_SELECTOR_RE.captures(content) {
        Some(cap) => extract_label_pairs(&cap[1]),
        None => FxHashMap::default(),
    }
}

fn build_selector_edges(
    frag: &Fragment,
    pods_with_labels: &[(FragmentId, FxHashMap<String, String>)],
    edges: &mut EdgeDict,
) {
    let (kind, _) = extract_resource_info(&frag.content);
    if kind.as_deref() != Some("Service") {
        return;
    }

    let selector = get_service_selector(&frag.content);
    if selector.is_empty() {
        return;
    }

    for (pod_id, labels) in pods_with_labels {
        if *pod_id != frag.id && labels_match(&selector, labels) {
            add_edge(edges, &frag.id, pod_id, SELECTOR_WEIGHT, REVERSE_FACTOR);
        }
    }
}

fn build_image_edges(
    frag: &Fragment,
    images: &FxHashMap<String, Vec<FragmentId>>,
    edges: &mut EdgeDict,
) {
    for cap in IMAGE_RE.captures_iter(&frag.content) {
        let image = cap[1].trim();
        if image.is_empty() || image.starts_with('$') {
            continue;
        }
        if let Some(other_ids) = images.get(image) {
            for other_id in other_ids {
                if *other_id != frag.id {
                    add_edge(edges, &frag.id, other_id, IMAGE_WEIGHT, REVERSE_FACTOR);
                }
            }
        }
    }
}

pub struct KubernetesEdgeBuilder;

impl EdgeBuilder for KubernetesEdgeBuilder {
    fn build(&self, fragments: &[Fragment], _repo_root: Option<&Path>) -> EdgeDict {
        let k8s_fragments: Vec<&Fragment> = fragments
            .iter()
            .filter(|f| is_kubernetes_manifest(Path::new(f.path()), &f.content))
            .collect();

        if k8s_fragments.is_empty() {
            return EdgeDict::default();
        }

        let mut edges = EdgeDict::default();
        let idx = build_resource_index(&k8s_fragments);

        for frag in &k8s_fragments {
            build_configmap_edges(frag, &idx.configmaps, &mut edges);
            build_secret_edges(frag, &idx.secrets, &mut edges);
            build_service_edges(frag, &idx.services, &mut edges);
            build_volume_edges(frag, &idx.pvcs, &mut edges);
            build_selector_edges(frag, &idx.pods_with_labels, &mut edges);
            build_image_edges(frag, &idx.images, &mut edges);
        }

        edges
    }

    fn discover_related_files(
        &self,
        changed: &[PathBuf],
        candidates: &[PathBuf],
        _repo_root: Option<&Path>,
        file_cache: Option<&FxHashMap<PathBuf, String>>,
    ) -> Vec<PathBuf> {
        let k8s_changed: Vec<&PathBuf> = changed
            .iter()
            .filter(|p| {
                if !is_yaml_file(p) {
                    return false;
                }
                match base::read_file_cached(p, file_cache) {
                    Some(content) => is_kubernetes_manifest(p, &content),
                    None => false,
                }
            })
            .collect();

        if k8s_changed.is_empty() {
            return vec![];
        }

        let k8s_dirs = collect_k8s_dirs(&k8s_changed);
        let changed_set: FxHashSet<&PathBuf> = changed.iter().collect();
        let mut discovered = Vec::new();

        for candidate in candidates {
            if changed_set.contains(candidate) || !is_yaml_file(candidate) {
                continue;
            }
            if !is_in_k8s_dir(candidate, &k8s_dirs) {
                continue;
            }
            if let Some(content) = base::read_file_cached(candidate, file_cache) {
                if is_kubernetes_manifest(candidate, &content) {
                    discovered.push(candidate.clone());
                }
            }
        }

        discovered
    }
}
